agent/tool-scripts/pprof

```diff
@@ -101,88 +102,75 @@ done
+            master_pids[${master}_api]=$!
+            oc get --raw /debug/pprof/heap --server https://${master}:8443 > ${master}_api_heap && ${tool_bin} tool ${tool} -text -seconds=${interval} -output $PPROF_TMPDIR/${master}/$(date +%Y-%m-%d:%H:%M:%S)_api_heap.txt /tmp/${master}_api_heap &
+            oc get --raw /debug/pprof/profile?seconds=${interval} --server https://${master}:8444 > ${master}_controller_profile && ${tool_bin} tool ${tool} -text -seconds=${interval} -output $PPROF_TMPDIR/${master}/$(date +%Y-%m-%d:%H:%M:%S)_controller_profile.txt /tmp/${master}_controller_profile &
+            master_pids[${master}_controller]=$!
+            oc get --raw /debug/pprof/heap --server https://${master}:8444 > ${master}_controller_heap && ${tool_bin} tool ${tool} -text -seconds=${interval} -output $PPROF_TMPDIR/${master}/$(date +%Y-%m-%d:%H:%M:%S)_controller_heap.txt /tmp/${master}_controller_heap &
+            if (( i++%${#masters[@]}==0 )); then
+		for pid in ${master_pids[*]}; do
+                    wait ${pid}
+                done
+            fi
+        done
+    done
 }
 
-collect_data() {
+get_hosts() {
+    declare -A node_pids
     while true; do
-        if [ $(systemctl list-unit-files | grep atomic-openshift-master.service | awk {'print $2'}) == "masked" ]; then
-            $tool_bin tool $tool -text -seconds=$interval -output $PPROF_TMPDIR/$(date +%Y-%m-%d:%H:%M:%S)_controller.txt http://localhost:6060/debug/pprof/profile 
-            $tool_bin tool $tool -text -seconds=$interval -output $PPROF_TMPDIR/$(date +%Y-%m-%d:%H:%M:%S)_controller_heap.txt http://localhost:6060/debug/pprof/heap 
-            $tool_bin tool $tool -text -seconds=$interval -output $PPROF_TMPDIR/$(date +%Y-%m-%d:%H:%M:%S)_api.txt http://localhost:6061/debug/pprof/profile
-            $tool_bin tool $tool -text -seconds=$interval -output $PPROF_TMPDIR/$(date +%Y-%m-%d:%H:%M:%S)_api_heap.txt http://localhost:6061/debug/pprof/heap 
-        else
-            $tool_bin tool $tool -text -seconds=$interval -output $PPROF_TMPDIR/$(date +%Y-%m-%d:%H:%M:%S).txt http://localhost:6060/debug/pprof/profile 
-            $tool_bin tool $tool -text -seconds=$interval -output $PPROF_TMPDIR/$(date +%Y-%m-%d:%H:%M:%S)_heap.txt http://localhost:6060/debug/pprof/heap 
-        fi
-    done 
-} 
+        j=1
+        for host in "${hosts[@]}"; do
+            oc get --raw /debug/pprof/profile?seconds=${interval} --server https://${host}:10250 > ${host}_profile && ${tool_bin} tool ${tool} -text -seconds=${interval} -output $PPROF_TMPDIR/${host}/$(date +%Y-%m-%d:%H:%M:%S)_profile.txt /tmp/${host}_profile &
+            node_pids[${host}]=$!
+            oc get --raw /debug/pprof/heap --server https://${host}:10250 > ${host}_heap && ${tool_bin} tool ${tool} -text -seconds=${interval} -output $PPROF_TMPDIR/${host}/$(date +%Y-%m-%d:%H:%M:%S)_heap.txt /tmp/${host}_heap &
+            if (( j++%${#hosts[@]}==0 )); then
```

Comment:
Why is this check inside the for loop?  Shouldn't this be done after line 147?

Reply:
Why would it be outside of the loop? We need to iterate `j` and see what value it's at?

Reply:
Perhaps we need to be clear what is happening in the `if` statement?

It seems to be:

 1. increment `j` by one (inline)
 2. find the modulus of the incremented value of `j` with the length of hosts
 3. compare that modulus value to `0`
 4. if the value is `0`, then proceed to issue the `for` loop below (to wait for all node pids)

If that is true, won't that `for` loop only be executed when the last host has been processed?  If so, isn't that the same as just executing the "wait" `for` loop after the `host` for loop?

If that is true, in order to make this code a bit more supportable, it seems reasonable to move the wait `for` loop outside of the host `for` loop and remove the somewhat arcane bash syntax for and post-increment logic.

Reply:
You are correct, I had added it for some reason a the time but it is no longer clear to me why. With the for loop over the masters/hosts it does seem redundant.

Reply:
Changed in https://github.com/distributed-system-analysis/pbench/pull/847